cloudferrylib/os/actions/cinder_database_manipulation.py

```diff
@@ -47,18 +110,314 @@ def run(self, *args, **kwargs):
+
+        for try_dst in dst:
+            try:
+                self._run_rsync(src, try_dst)
+                return try_dst
+            except RemoteExecutionError:
+                pass
+        raise RemoteExecutionError('Copy to %s failed.', dst)
+
+    def _mount_output(self, cloud, vt=None):
+        if vt:
+            cmd = (
+                "for exp in "
+                r"$(awk -F= '/\["
+                "%s"
+                r"\]/{i=1} /!/ {i=0} i && /nfs_shares_config/ "
+                "{print $2; exit}' /etc/cinder/cinder.conf | xargs cat); "
+                "do mount | "
+                "awk '{if (match($1, \"^'$exp'$\") && $3 ~ \"cinder\") "
+                "{print $3\"%s\"$1}}'; done"
+            ) % (vt['name'], MOUNT_DELIM)
+        else:
+            # default nfs_shares_config
+            cmd = (
+                "for exp in "
+                "$(awk -F= '/nfs_shares_config/ "
+                "{print $2; exit}' /etc/cinder/cinder.conf | xargs cat); "
+                "do mount | "
+                "awk '{if (match($1, \"^'$exp'$\") && $3 ~ \"cinder\") "
+                "{print $3\"%s\"$1}}'; done"
+            ) % MOUNT_DELIM
+        res = self._run_cmd(cloud, cmd)
+        return res if isinstance(res, list) else [res]
+
+    def mount_dirs(self, cloud, vt=None):
+        """
+        Get shares from mount output.
+
+        :return: list of paths
+
+        """
+        if vt:
+            cmd = (
+                "for exp in "
+                r"$(awk -F= '/\["
+                "%s"
+                r"\]/{i=1} /!/ {i=0} i && /nfs_shares_config/ "
+                "{print $2; exit}' /etc/cinder/cinder.conf | xargs cat); "
+                "do mount | "
+                "awk '{if (match($1, \"^'$exp'$\") && $3 ~ \"cinder\") "
+                "{print $3}}'; done"
+            ) % vt['name']
+        else:
+            # default nfs_shares_config
+            cmd = (
+                "for exp in "
+                "$(awk -F= '/nfs_shares_config/ "
+                "{print $2; exit}' /etc/cinder/cinder.conf | xargs cat); "
+                "do mount | "
+                "awk '{if (match($1, \"^'$exp'$\") && $3 ~ \"cinder\") "
+                "{print $3}}'; done"
+            )
+        res = self._run_cmd(cloud, cmd)
+        return res if isinstance(res, list) else [res]
+
+    def _vt_map(self):
+        # host volume_type_id->hostname map
+        # cached property
+        if not hasattr(self, 'dst_volume_types') \
+                or self.dst_volume_types is None:
+            self.dst_volume_types = _volume_types_map(self.dst_data)
+        res = dict(
+            [(vt['id'], self.dst_volume_types[vt['name']])
+             for vt in _volume_types(self.data)
+             if vt['name'] in self.dst_volume_types])
+        return res
+
+    def _dst_host(self, vtid=None):
+        # vtid -> dst_host
+        # cached property
+        if not hasattr(self, 'dst_hosts') or self.dst_hosts is None:
```

Comment:
dst_host is always initialized in the __init__ function, "self.dst_hosts is None" is enough.